pkg/util/log/sampled.go

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+type sampledError struct {
+	err     error
+	sampler *Sampler
+}
+
+func (s sampledError) Error() string {
+	if s.sampler == nil {
+		return s.err.Error()
+	}
+	return fmt.Sprintf("%s (sampled 1/%d)", s.err.Error(), s.sampler.freq)
+}
+
+func (s sampledError) Unwrap() error { return s.err }
+
+// ShouldLog is called by common logging module.
+func (s sampledError) ShouldLog(_ context.Context, _ time.Duration) bool {
+	return s.sampler == nil || s.sampler.Sample()
+}
+
+type Sampler struct {
+	freq  int64
+	count atomic.Int64
+}
+
+func newSampler(freq int64) *Sampler {
+	if freq == 0 {
+		return nil
+	}
+	return &Sampler{freq: freq}
+}
+
+func (s *Sampler) Sample() bool {
+	count := s.count.Inc()
+	return (count-1)%s.freq == 0
+}
+
+type Samplers struct {
+	freq     int64
+	samplers map[string]*Sampler
+}
+
+func NewSamplers(freq int64) *Samplers {
+	if freq == 0 {
+		return nil
+	}
+	samplers := make(map[string]*Sampler)
+	return &Samplers{
+		freq:     freq,
+		samplers: samplers,
+	}
+}
+
+func (s *Samplers) sampler(err error) *Sampler {
+	sampler, ok := s.samplers[err.Error()]
+	if !ok {
+		sampler = newSampler(s.freq)
+		s.samplers[err.Error()] = sampler
+	}
```

Comment:
This needs to be synchronized. And I don't think we want to introduce a mutex here, so we should come up with a way of initializing all samples from the beginning.